JOSS Submission (#622)
PlasmaFAIR/fortitude: joss/paper.md

---
title: 'Fortitude: The Fortran Linter'
tags:
  - Fortran
  - static analysis
  - linting
  - linter
  - tool
  - tools
  - tooling
authors:
  - name: Liam Pattinson
    orcid: 0000-0001-8604-6904
    equal-contrib: true
    corresponding: true
    affiliation: 1
  - name: Peter Hill
    orcid: 0000-0003-3092-1858
    equal-contrib: true
    affiliation: 1
  - name: Ian McInerney
    orcid: 0000-0003-2616-9771
    affiliation: 2
  - name: Connor Aird
    affiliation: 3
affiliations:
 - name: University of York, United Kingdom
   index: 1
 - name: Imperial College London, United Kingdom
   index: 2
 - name: University College London, United Kingdom
   index: 3
date: 14 April 2026
bibliography: paper.bib
---

# Summary

Static analysis tools are used by software developers to better understand and
improve their software, and they achieve this by examining their source code
and/or binaries without running the software. A ‘linter’ is a type of static
analysis tool that offers an opinionated critique of software beyond its
syntactic validity, highlighting bug-prone coding patterns, deviations from

> common style guides, and ways to upgrade to more modern features.

Reviewer: Not necessarily more modern, but generally "better" ways?

```suggestion
common style guides, and generally, suggestions to improve the code in some way.
```

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/82a4028d29f8962e2dff1c0423f40f3a557431d1

Fortitude is a linter targeting Fortran, which has hitherto lacked a standard
open-source solution. It is orders of magnitude faster than its open-source
counterparts, is highly customisable, has robust parsing capabilities, can

> automatically fix linting issues, and can integrate directly into a user’s

Reviewer 2: I feel like the autofix functionality could be featured more prominently, perhaps in the section when comparing with other linters (I am not sure if they offer it as well or not)?. To me this is the killer feature that makes introducing a linter much easier, even to people who would otherwise not be open to the idea.

Author: Thanks, that's a good point!

Author: Fixed in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/99977051f6f72607fa1acd975966bc51a4515d1b

> editor or IDE.

Reviewer: ```suggestion
Fortitude is a linter targeting Fortran, which has hitherto lacked a standard
open-source solution. It is two orders of magnitude faster than its open-source
counterparts, is highly customisable, has robust parsing capabilities, can
automatically fix many linting issues, and can integrate directly into a user’s
editor or IDE.
```

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/82a4028d29f8962e2dff1c0423f40f3a557431d1

> editor or IDE.

Reviewer 3: ```suggestion
automatically fix linting issues, and can integrate either into a CI/CD pipeline
or directly into a user’s editor or IDE.
```

I think it's good to say this can be used in CI/CD and isn't just an editor plugin.

# Statement of need

Fortran is the oldest programming language still in common usage today, and is
widely used for high-performance research applications in fields such as

> climatology, quantum condensed matter physics, and fusion energy. The longevity

Reviewer: We could probably add citations here. Probably also worth referencing the Archer report that had like 60% of the cycles used by Fortran

Author: Any idea exactly what to cite for Fortran?

Reviewer 3: Not directly Archer, but there was a recent article about "The state of Fortran" https://doi.org/10.1109/MCSE.2022.3159862 and this HPC position paper (https://doi.org/10.1177/10943420231166608) has HPC Myth 10 "Fortran is dead, long live the DSL" which talks about how Fortran is still relevant.

Author: Thanks @imciner2, that's really helpful!

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/f8d6c820509fd3079a17acd6d574cd9c3ce2bc9c and https://github.com/PlasmaFAIR/fortitude/pull/622/commits/955f3c320da143960cda56c68823ea3224a94e6e

of Fortran means that much of the research software in use today is written to
older standards and contains a lot of technical debt. This places a large burden
on the maintainers of this research software, and porting Fortran codes to more
modern languages is not typically worth the resources required or the risks of
introducing new bugs or performance issues.

Fortitude is a Fortran linter designed to help writers and maintainers of
Fortran research software to improve the quality of their code. Its linting
rules are grouped under categories specifying what aspect of software they
intend to improve:

- Correctness: Rules to find bug-prone coding patterns, helping developers to
  catch errors early and improve the safety of their code.
- Obsolescent: Rules to flag Fortran features marked as obsolescent in the
  standard and recommend refactoring strategies to avoid them.
- Modernisation: Rules to update Fortran code to make use of newer features

>   beyond the recommendations of the standard.

Reviewer 3: I can't quite parse what you mean by the phrase "beyond the recommendations of the standard" - the idea of modernisation rules is to update code to use the newer features of the standard instead of older ones, so how is that beyond?

Author: I probably should be more specific here. The obsolescent rules are for things specifically marked 'obsolescent'/'deprecated' in the Fortran standards, while 'modernisation' is for things that aren't strictly obsolescent but there's a nicer way of doing things.

Author: Fixed in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/a3ad385168b2b8038b8f62ea8401fefb2cc5f0f1 (kinda)

- Style: Rules to make Fortran code more readable and help adhere to a common
  set of coding conventions.
- Portability: Rules to avoid compiler- or platform-specific features in favour
  of portable alternatives.

As an example, the correctness rule `nonportable-shortcircuit-inquiry` detects a
subtle bug in which an optional argument is both checked and used within the
same logical expression, e.g:

```f90
if (present(arg) .and. arg > 0) then
    print *, arg
end if
```

In languages like C and Python, the order of operations in a logical expression

> is guaranteed to run left-to-right, so equivalent statements would be safe. A

Reviewer: It's the combination of this and the lack of shortcircuiting that causes the bug

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/82a4028d29f8962e2dff1c0423f40f3a557431d1

programmer may expect similar behaviour in Fortran, but as this is not

> guaranteed by the standard, it is possible that this could result in a reference

Reviewer 3: Perhaps mention that since it is not in the standard, it is compiler-dependent so code that works on one compiler/platform might not work on another (hindering portability).

Author: Fixed in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/a3ad385168b2b8038b8f62ea8401fefb2cc5f0f1

to invalid data and a critical error. Running Fortitude over this code with the
rule activated delivers a diagnostic message to the user:

```console
test.f90:12:32: C161 variable inquiry `present(arg)` and use in same logical expression
   |
12 |         if (present(arg) .and. arg > 0) then
   |                                ^^^ C161
13 |             print *, arg
14 |         end if
   |

fortitude: 1 files scanned.
Number of errors: 1
```

> The use of a linter is especially beneficial when working on large projects with

Reviewer 2: I really like the example of the correctness rule (had no idea! 😱 ). I think it would be great to include one example from `modernisation` and / or `style`,categories especially for readers who are not familiar with linters (and who might not even no what a  "style" means in the programming context :-) ). This would also help to answer the question of @imciner2 above regarding the definition of the `modernisation` rules.

Author: I'll have a think about adding another example, but I'm way of the tight word limit on JOSS papers. A style rule might be a good idea as a lot of them are quite easy to explain and have autofixes.

Author: Fixed in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/a3ad385168b2b8038b8f62ea8401fefb2cc5f0f1

multiple developers, as it allows the team to enforce a consistent coding style.
To aid in this, the set of linting rules and other customisations may be set in
a configuration file and saved at a project level. Fortitude also offers a
pre-commit hook so teams can ensure that code is passing all checks before being
committed to a repository.

Fortitude may be used alongside other tooling in the Fortran ecosystem. For
example, its settings may be specified in the `fpm.toml` files used by the
Fortran Package Manager (FPM) [@fpm]. Fortitude’s editor integration using the
Language Server Protocol (LSP), by which it can provide suggestions inline with
the user’s code, is compatible with that of FortLS [@fortls], a plugin that can

> aid users in navigating a project. It can also be used alongside the fprettify

Reviewer 2: Oh, I use both `fortitude` and `fprettify` in my project, glad that this is an "officially" tested combination!

Author: Well, I wouldn't say we're strictly testing it, but avoiding clashes with fprettify is often in the back of my mind when I'm working on style rules. When we finally add a formatting mode, my aim would be to replicate fprettify's style in the first instance.

[@fprettify] code formatter without generating conflicts.

# State of the field

Fortitude was developed due to notable deficiencies in existing open-source
Fortran linting tools, most of which are unmaintained, lacking in features, use

> unreliable Fortran parsers, or have poor performance. The decision to write a

Reviewer: ```suggestion
Fortran linting tools, most of which are some combination of unmaintained, lacking in features, use
unreliable Fortran parsers, and/or have poor performance. The decision to write a
```

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/82a4028d29f8962e2dff1c0423f40f3a557431d1

new linter rather than contribute to existing solutions was inspired by the much
higher quality of linters in other languages; it was easier to adapt those
solutions to work with Fortran than to upgrade the existing Fortran linters to
meet the state-of-the-art. Fortitude borrows many language-agnostic assets from
the Python linter Ruff [@ruff], including much of the command-line interface,
output formats, and configuration file design. While these structural elements
could be reused directly, the linting rules themselves had to be written
specifically for Fortran.

Other popular open-source Fortran linting tools include:

- CamFort: Primarily a code refactoring tool, it features some linting rules
  [@Orchard:2013]. Written in Haskell, it features its own Fortran parser.
- Flint: A linter written in Python. Uses lizard to analyse cyclomatic
  complexity and related metrics, and regular expressions otherwise [@flint].
- Stylist: A linter written in Python and using FortranParser to generate an
  AST [@stylist].

To compare Fortitude’s performance to these tools, each was used to lint 72
files in the GS2 source code, a turbulence modelling tool used in

> magnetically-confined fusion energy research [@gs2]. Each tool was set up as

Reviewer 2: Should some basic info about the benchmark machine be included? (not sure if necessary)

Author: Fixed in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/97aa67183dfaedbdc70dd0a90923f1073e1f2d02. I'm not sure how much detail is needed, but hopefully the gist of "this is on an unremarkable laptop" is carried across.

follows:

- Fortitude v0.8.0: Activated all 87 rules.
- Stylist v0.5.dev (latest on GitHub): Activated all 11 rules.
- CamFort v1.3.dev (latest on GitHub): Activated the 5 rules in the
  `basic-checks` command.
- Flint v0.7.1: Activated the default set of 38 rules. One file had to be
  excluded due to a parse error.

The runtime of each tool is shown in \autoref{fig:performance}. Despite
implementing many more checks than the alternatives, Fortitude was found to run
between 50 and 100 times as quickly than its main competitors.

![Time taken for Fortran linters to lint 72 files in the GS2 project.\label{fig:performance}](performance_plot.pdf)

Further work is needed to bring Fortitude up to parity with the most popular
linting tools in other languages. For instance, despite borrowing heavily from
Ruff [@ruff], Fortitude does not implement a code formatting mode, and Ruff

> implements over 900 rules compared to the 87 rules in Fortitude v0.8.0. It also

Reviewer: This is maybe an unfair comparison (on us!), given that Ruff has some rules for specific libraries, and maybe we don't expect different languages to have the same number of linting rules.

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/82a4028d29f8962e2dff1c0423f40f3a557431d1

lacks some advanced checks that linters based on compilers can perform due to
their ability to access more semantic information within a project, such as
clang-tidy for C++ [@clangtidy].

# Software design

As linting is effectively a solved problem in other languages, the design of
Fortitude was guided by the reuse of existing resources. The core architecture
and user interface of Fortitude was inspired primarily by Ruff [@ruff], a
widely-used Python linter, and, where possible, Ruff’s language-agnostic

> features were directly repurposed. This reuse of a tried-and-tested design

Reviewer: Maybe mention this is thanks to it being open-source?

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/82a4028d29f8962e2dff1c0423f40f3a557431d1

permitted much more rapid development than would otherwise be possible, and
resulted in a user experience that can be readily understood by programmers who
are already familiar with Ruff.

However, the core checking loop of Fortitude differs to that of Ruff. Most rules
that operate over the Concrete Syntax Tree (CST) inherit a common trait that
requires them to specify their 'entrypoints': the CST node types on which they
should be activated. When Fortitude scans the CST of any Fortran files, it
checks whether the user has requested the activation of any rules that start on
each node type it encounters, and runs each of these checks in turn. This way,

> Fortitude only needs to perform a single pass over the CST for all activated

Reviewer 2: Is the fact that this "single-pass" design works a consequence of Fortran language structure? In other words, why doesn't ruff do it this way as well?
(disclaimer: I have no idea what I am talking about here, just thought it might interesting to explain this difference a bit more).

Author: I believe Ruff works similarly in that it performs a single pass over the AST/CST, but it doesn't have anything like our 'entrypoints' system. Instead, it has some [very long functions](https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/src/checkers/ast/analyze/expression.rs) that essentially do the same thing but in a way that's much harder to maintain.

We're discovering that there are some benefits to Ruff's method, which means we'll likely adopt a hybrid approach in future (e.g. if a group of rules have some shared processing to do beforehand, it's much faster to do it in one place and pass the results to each checking function).

rules, rather than performing a full pass per rule. This is largely how it

> achieves such high performance.

Reviewer: This is not super clear about the difference between Fortitude and Ruff. Really, the main difference here is that Ruff uses a hard coded visitor pattern, while Fortitude's is dynamically generated via the entrypoints

Author: This paragraph was a last-minute addition after I read the [JOSS example paper](https://joss.readthedocs.io/en/latest/example_paper.html), which goes into a surprising amount of detail on software design. It goes into details about which base classes exist and what kinds of data structures are used. I personally would have expected that level of detail to be too much for JOSS, and the entrypoints system is probably the closest thing we have to that.

If you think we should be okay with a higher-level description of the software design thought process, I'd lean towards just removing this paragraph. Otherwise, I'll see if I can clean it up a bit.

given the high degree of backwards compatibility in the Fortran standards.
Rather than writing a Fortran parser from scratch, this was achieved using the
TreeSitter parsing framework, which provides a fast and robust solution with
Rust bindings [@treesitter]. The Fortran extension, tree-sitter-fortran, has
itself received numerous upgrades thanks to the experience gained working on
Fortitude [@treesitterfortran].

# Research impact statement

Fortitude began life as an in-house tool to aid in the refactoring of legacy
Fortran codes within plasma physics, and is now used for code quality control by
teams all over the world, with users ranging from individual researchers to
government institutions. The earliest known external use was for FTorch
[@Atkinson:2025], a library for running neural network models in Fortran. The
Met Office uses Fortitude for various projects, including LFRic [@lfric], CASIM
[@casim], and Socrates [@socrates]. It is also used by the quantum chemistry and
solid state physics package cp2k [@cp2k] and the stellar astrophysics package
MESA [@Paxton:2011].

The number of contributors to Fortitude has grown from the original two authors

> to over 19, and more community members have raised feature requests and bug

Reviewer: Blimey, I didn't realise we had this many contributors!

Author: It's actually more than 19, for some reason it isn't yet counting some of the latest merged PRs

> reports. It receives over 4,000 downloads per month via PyPI, and a further

Reviewer: Pypistats is very unreliable, this number includes automated mirrors and CI etc

Author: I agree, this was another case of following the lead from the [JOSS example paper](https://joss.readthedocs.io/en/latest/example_paper.html). I'm more inclined to trust GitHub stars as an indication of users, but wouldn't know how to estimate the ratio of users per star.

GitHub's stats indicate 73 unique visitors to our repo in the last 2 weeks, but again, I don't know how that translates to active users. I imagine a lot of people started using Fortitude a while back and haven't checked in on it in a while.

unknown number of downloads of platform-specific binaries using a provided
installer script. With over 190 GitHub stars, Fortitude appears to be the most
popular open-source Fortran linting tool currently in use.

# AI usage disclosure

Some of the code in Fortitude was created with the assistance of the generative
AI tool GitHub Copilot, but none has been generated wholesale, and all code has
been verified by at least two human developers prior to deployment. The software
is thoroughly tested, and all tests have been designed manually. Besides
spelling and grammar checking, this paper was written without AI assistance.

# Acknowledgements

> Fortitude's development has been funded by the PlasmaFAIR project.

Reviewer: ```suggestion
Fortitude's development has been funded by the PlasmaFAIR project, EPSRC Grant EP/V051822/1.
```

Author: Thanks, I was meant to ask you about this and completely forgot!

analysis tool that offers an opinionated critique of software beyond its
syntactic validity, highlighting bug-prone coding patterns, deviations from
common style guides, use of outdated features, and, generally, suggestions to
improve the code in some way.

Fortitude is a linter targeting Fortran, which has hitherto lacked a standard
open-source solution. It is two orders of magnitude faster than its open-source
counterparts, is highly customisable, has robust parsing capabilities, can
automatically fix many linting issues, and can integrate either into a CI/CD
pipeline or directly into a user’s editor or IDE.

# Statement of need

The Fortran programming language, first released in 1957 [@fortran:1957], is the
oldest high-level programming language still in common usage
today [@fortran:2022], and is widely used for high-performance research
applications in fields such as climatology, quantum condensed matter physics,
and fusion energy. Fortran powers the majority of the software run on Archer2,
which until 2024 was the most powerful supercomputer in the United
Kingdom [@archer2].

The longevity of Fortran means that much of the research software in use today

> is written to older standards and contains a lot of technical debt. This places

Reviewer 4: 3 words

```suggestion
Due to Fortran's logenvity, much of the research software in use today
is written to older standards containing a lot of technical debt. This places
```

a large burden on the maintainers of this research software, and porting Fortran
codes to more modern languages is not typically worth the resources required or
the risks of introducing new bugs or performance issues.

Fortitude is a Fortran linter designed to help writers and maintainers of

> Fortran research software to improve the quality of their code. Its linting

Reviewer 4: 3 words (6 total)
```suggestion
Fortitude is a Fortran linter designed to help developers of
Fortran research software improve the quality of their code. Its linting
```

rules are grouped under categories specifying what aspect of software they
intend to improve:

- Correctness: Rules to find bug-prone coding patterns, helping developers to

>   catch errors early and improve the safety of their code.

Reviewer 4: 1 words (7 total)
```suggestion
- Correctness: Rules to find bug-prone coding patterns, helping developers
  catch errors early and improve the safety of their code.
```

- Obsolescent: Rules to flag features marked as obsolescent in the Fortran
  standard and recommend refactoring strategies to avoid them.

> - Modernisation: Rules to update Fortran code to make use of newer features.

Reviewer 4: 2 words (9 total)
```suggestion
- Modernisation: Rules to update Fortran code to utilise newer features.
```

  These are complementary to 'obsolescent' rules, and go beyond the strict
  recommendations of the Fortran standard.
- Style: Rules to make Fortran code more readable and help adhere to a common
  set of coding conventions.
- Portability: Rules to avoid compiler- or platform-specific features in favour
  of portable alternatives.

As an example, the correctness rule `nonportable-shortcircuit-inquiry` detects a
subtle bug in which an optional argument is both checked and used within the
same logical expression, e.g:

```f90
if (present(arg) .and. arg > 0) then
    print *, arg
end if
```

In languages like C and Python, the order of operations in a logical expression
is guaranteed to run left-to-right and will exit as soon as the outcome of the
expression is known, skipping any remaining operations. A programmer may
expect similar behaviour in Fortran, but as this is compiler-dependent behaviour
in Fortran and not guaranteed by the standard, it is possible that this could
result in a reference to invalid data and a critical error. Running Fortitude
over this code with the rule activated delivers a diagnostic message to the

> user:

Reviewer 4: 10 words (19 total)
```suggestion
Unlike in languages such as C and Python, the order of operations in a logical
Fortran expression is not guaranteed to run left-to-right and may result in all
operations within the expression being executed. As this is not guaranteed by
the Fortran standard, it is compiler-dependent behaviour. Therefore, it is possible
that this could result in a reference to invalid data and a critical error. Running
Fortitude over this code with the rule activated delivers a diagnostic message to
the user:
```

  excluded due to a parse error.

The runtime of each tool is shown in \autoref{fig:performance}. In all cases,
the tests were run on a simple laptop featuring a 4-core 11th generation Intel
i5 CPU running at 2.60GHz. Despite implementing many more checks than the
alternatives, Fortitude was found to run between 50 and 100 times as quickly
than its main competitors.

![Time taken for Fortran linters to lint 72 files in the GS2 project.\label{fig:performance}](performance_plot.pdf)

# Software design

As linting is effectively a solved problem in other languages, the design of
Fortitude was guided by the reuse of existing resources. The core architecture
and user interface of Fortitude was inspired primarily by Ruff [@ruff], a
widely-used Python linter, and, where possible, Ruff’s language-agnostic
features were able to be directly repurposed thanks to its open and permissive
licensing. This reuse of a tried-and-tested design enabled much more rapid
development than would otherwise be possible, and has resulted in a user
experience that can be readily understood by programmers who are already
familiar with Ruff.

> The generation of the CST from a Fortran file is no simple task, especially

Reviewer 4: Have we defined CST?

given the high degree of backwards compatibility in the Fortran standards.
Rather than writing a Fortran parser from scratch, this was achieved using the
TreeSitter parsing framework, which provides a fast and robust solution with
Rust bindings [@treesitter]. The Fortran extension, tree-sitter-fortran, has
itself received numerous upgrades thanks to the experience gained working on
Fortitude [@treesitterfortran].

> 

Reviewer 4: 6 words (25 total)
```suggestion
Therefore, this was achieved using the pre-existing TreeSitter parsing framework,
which provides a fast and robust solution with Rust bindings [@treesitter]. The Fortran
extension, tree-sitter-fortran, has itself received numerous upgrades thanks to the
experience gained working on Fortitude [@treesitterfortran].

```

# Research impact statement

> Fortitude began life as an in-house tool to aid in the refactoring of legacy

Reviewer 4: 1 words (26 total)
```suggestion
Fortitude began as an in-house tool to aid in the refactoring of legacy
```

    affiliation: 4
  - name: Violet Sherratt
    orcid: 0009-0009-0066-8338
    affiliation: 5
affiliations:
  - name: University of York, United Kingdom
    index: 1
  - name: University College London, United Kingdom
    index: 2
  - name: INRAE (French National Research Institute for Agriculture, Food and Environment), France
    index: 3
  - name: Imperial College London, United Kingdom
    index: 4
  - name: Met Office, United Kingdom
    index: 5
date: 21 April 2026
bibliography: paper.bib

rsecon26: accepted
---

# Summary

Static analysis tools are used by software developers to better understand and
improve their software, and they achieve this by examining their source code
and/or binaries without running the software. A ‘linter’ is a type of static
analysis tool that offers an opinionated critique of software beyond its
syntactic validity, highlighting bug-prone coding patterns, deviations from
common style guides, use of outdated features, and, generally, suggestions to
improve the code in some way.

Fortitude is a linter targeting Fortran, which has hitherto lacked a standard
open-source solution. It is over an order of magnitude faster than its open-source

> counterparts, is highly customisable, has robust parsing capabilities, can

Reviewer 5: I think a possible reading of "faster than its open-source counterparts" is that fortitude itself is *not* open-source. I think it's specifically the word "counterpart" that gives that impression?

```suggestion
open-source solution. It is over an order of magnitude faster than other open-source
offerings, is highly customisable, has robust parsing capabilities, can
```

Author: I think you make a good point. I've just made the v0.9.1 release that will accompany the JOSS publication, but as it turns out, I may have missed another co-author from the list, so I'll need to make a v0.9.2 release just to correct that. If they take up the offer of co-authorship, I'll also get this change in.

The Fortran programming language, first released in 1957 [@fortran:1957], is the
oldest high-level programming language still in common usage
today [@fortran:2022], and is widely used for high-performance research
applications in fields such as climatology, quantum condensed matter physics,
and fusion energy. Fortran powers the majority of the software run on Archer2,
which until 2024 was the most powerful supercomputer in the United
Kingdom [@archer2].

Due to Fortran's longevity, much of the research software in use today
is written to older standards and contains a lot of technical debt. This places
a large burden on the maintainers of this research software, and porting Fortran
codes to more modern languages is not typically worth the resources required or
the risks of introducing new bugs or performance issues.

Fortitude is a Fortran linter designed to help developers of Fortran research
software improve the quality of their code. Its linting rules are grouped
under categories specifying what aspect of software they intend to improve:

- Correctness: Rules to find bug-prone coding patterns, helping developers
  catch errors early and improve the safety of their code.

> - Obsolescent: Rules to flag features marked as obsolescent in the Fortran

Reviewer 5: This appears to be the only rule group name that is not a noun - surely it would be more consistent to call it "obsolescence"?

I note however that this is not just a feature of this paper, "obsolescent" is consistently used over "obsolescence" throughout the repo, so this is probably an entirely separate discussion. But I thought I'd still raise it here as I hadn't even noticed the pattern before seeing it written out in paper form.

Author: That's... a very good point. I'm a bit shocked that I hadn't noticed this all this time.

We have a mechanism to redirect outdated rule-codes, and this should work just fine for ensuring `--select=obsolescent` and `--select=obsolescence` have the same outcome. We could make this update in time for v0.10.0, which should be coming out soon, but unfortunately I think that would be too much of a change for the paper at this stage.

---
title: 'Fortitude: The Fortran Linter'
tags:
  - Fortran
  - static analysis
  - linting
  - linter
  - tool
  - tools
  - tooling
authors:
  - name: Liam Pattinson
    orcid: 0000-0001-8604-6904
    equal-contrib: true
    corresponding: true
    affiliation: 1
  - name: Peter Hill
    orcid: 0000-0003-3092-1858
    equal-contrib: true
    affiliation: 1
  - name: Ian McInerney
    orcid: 0000-0003-2616-9771

>     affiliation: 2
>   - name: Connor Aird
>     affiliation: 3

Reviewer 4: I am very happy to be included as a co-author
```suggestion
  - name: Connor Aird
    orcid: 0009-0001-7879-3472  
    affiliation: 3
```

pipeline or directly into a user’s editor or IDE.

# Statement of need

The Fortran programming language, first released in 1957[@fortran:1957], is the
oldest high-level programming language still in common usage
today[@fortran:2022], and is widely used for high-performance research
applications in fields such as climatology, quantum condensed matter physics,
and fusion energy. Fortran powers the majority of the software run on Archer2,
which until 2024 was the most powerful supercomputer in the United
Kingdom[@archer2].

The longevity of Fortran means that much of the research software in use today
is written to older standards and contains a lot of technical debt. This places
a large burden on the maintainers of this research software, and porting Fortran
codes to more modern languages is not typically worth the resources required or
the risks of introducing new bugs or performance issues.

Fortitude is a Fortran linter designed to help writers and maintainers of
Fortran research software to improve the quality of their code. Its linting
rules are grouped under categories specifying what aspect of software they
intend to improve:

> - Correctness: Rules to find bug-prone coding patterns, helping developers to

Reviewer 4: I'm not sure if it is relevant for JOSS but you could include something about the lifecycle of a rule, i.e. that there is some sort of approval process and `--preview` state that ensure rules are stable before they are enforced by default.

Author: That's a good point, it shows that we're taking community feedback into account and trying to ensure the rules are robust before release. There have been a few occasions where we've had to rewrite rules between preview mode and the proper release.

Author: Resolved in https://github.com/PlasmaFAIR/fortitude/pull/622/commits/6d5513d410f8f97fd2d807ae3dadb0a41b9fc4e3

and/or binaries without running the software. A ‘linter’ is a type of static
analysis tool that offers an opinionated critique of software beyond its
syntactic validity, highlighting bug-prone coding patterns, deviations from
common style guides, use of outdated features, and, generally, suggestions to
improve the code in some way.

Fortitude is a linter targeting Fortran, which has hitherto lacked a standard
open-source solution. It is two orders of magnitude faster than its open-source
counterparts, is highly customisable, has robust parsing capabilities, can
automatically fix many linting issues, and can integrate either into a CI/CD
pipeline or directly into a user’s editor or IDE.

# Statement of need

The Fortran programming language, first released in 1957[@fortran:1957], is the
oldest high-level programming language still in common usage
today[@fortran:2022], and is widely used for high-performance research
applications in fields such as climatology, quantum condensed matter physics,
and fusion energy. Fortran powers the majority of the software run on Archer2,
which until 2024 was the most powerful supercomputer in the United

> Kingdom[@archer2].

Reviewer 3: ```suggestion
The Fortran programming language, first released in 1957 [@fortran:1957], is the
oldest high-level programming language still in common usage
today [@fortran:2022], and is widely used for high-performance research
applications in fields such as climatology, quantum condensed matter physics,
and fusion energy. Fortran powers the majority of the software run on Archer2,
which until 2024 was the most powerful supercomputer in the United
Kingdom [@archer2].
```
All other references have a space before the reference.

Author: Thanks, good catch!
